streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+                    currentStateTwo.put(row.key, row.value);
+                }
+            }
+
+            // phase 3:
+            // expected end state per output partition:
+            // p-0: 10 records + COMMIT + 5 records (pending)
+            // p-1: 10 records + COMMIT + 5 records (pending)
+            stateTransitions2.clear();
+            final AtomicBoolean errorInjectedClient2 = new AtomicBoolean(false);
+            errorInjected = errorInjectedClient2;
+            if (injectError) {
+                errorInjectedClient1.set(true);
+            }
+            final List<KeyValue<Long, Long>> uncommittedInputDataBeforeUpgrade = prepareData(10L, 15L, 0L, 1L);
+            allData.addAll(uncommittedInputDataBeforeUpgrade);
+
+            writeInputData(uncommittedInputDataBeforeUpgrade);
+
+            final List<KeyValue<Long, Long>> allOutputDataBeforeUpgrade = readResult(allData.size(), null);
+            checkResultPerKey(allOutputDataBeforeUpgrade, computeExpectedResult(allData, new HashMap<>()));
+
+            // phase 4:
+            // expected end state per output partition:
+            // stop case:
+            //   p-0: 10 records + COMMIT + 5 records + COMMIT
+            //   p-1: 10 records + COMMIT + 5 records (pending)
+            // crash case:
+            //   p-0: 10 records + COMMIT + 5 records (pending)
+            //   p-1: 10 records + COMMIT + 5 records (pending)
+            if (!injectError) {
+                streams1Alpha.close();
+            }
+            waitForCondition(
+                () -> stateTransitions2.equals(Arrays.asList(
+                    KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING)
+                )),
+                MAX_WAIT_TIME_MS,
+                "Client did not stabilize on time."
+            );
+            errorInjected = null;
+
+            final List<KeyValue<Long, Long>> allOutputDataAfterFirstUpgrade = readResult(allData.size(), null);
+            checkResultPerKey(allOutputDataAfterFirstUpgrade, computeExpectedResult(allData, new HashMap<>()));
+
+            if (!injectError) {
+                final List<KeyValue<Long, Long>> committedInputDataAfterFirstUpgrade =
+                    uncommittedInputDataBeforeUpgrade.stream().filter(pair -> pair.key == keyFilterFirstUpgrade).collect(Collectors.toList());
+
+                final List<KeyValue<Long, Long>> committedOutputDataAfterFirstUpgrade = readResult(committedInputDataAfterFirstUpgrade.size(), CONSUMER_GROUP_ID);
+                checkResultPerKey(committedOutputDataAfterFirstUpgrade, computeExpectedResult(committedInputDataAfterFirstUpgrade, currentStateOne));
+            }
+
+            // phase 5:
+            // expected end state per output partition:
+            // p-0: 10 records + COMMIT + 5 records + COMMIT
```

Comment:
Here p-0 may end with COMMIT (close) or ABORT (crash) right?

Reply:
Same comment as above: the is no pending TX for p-0 when a crash happens. But the comment here is still simplfied. The end state would be (in full details):
```
            // phase 5:
            // expected end state per output partition:
            // stop case (original task migrated back):
            //   p-0: 10 records + COMMIT + 5 records + COMMIT
            //   p-1: 10 records + COMMIT + 5 records (pending)
            // stop case (task "switch"):
            //   p-0: 10 records + COMMIT + 5 records + COMMIT
            //   p-1: 10 records + COMMIT + 5 records + COMMIT
            // crash case (original task migrates back):
            //   p-0: 10 records + COMMIT + 5 records + COMMIT
            //   p-1: 10 records + COMMIT + 5 records (pending)
            // crash case (task "switch"):
            //   p-0: 10 records + COMMIT + 5 records (pending)
            //   p-1: 10 records + COMMIT + 5 records + COMMIT
```

Reply:
I thought that before the client crashed there's a pending txn because we've processed 5 records already in stage 3):

```
3.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
        //      - both pending transactions are based on task producers
```

Is that not right?

Reply:
In the current implementation, we set the inject error flag to true, before we write the 5 input records. Hence, we crash when processing the first record -- therefore, no output record is written and no TX is started (as we start TX lazily).

We cannot set the inject error flag to true after we sent the input data, because we would be subject to a race condition -- all 5 records might have been processed before we set the flag and thus, so error would happen as `process()` won't be called any more. We can of course change the test to add an additional condition in the `Processor` to only throw an exception if the flag is set to true and if, eg, `value % 10 == 4` (ie, we throw when processing the last record, and thus get 4 pending writes).